src/Compilers/CSharp/Portable/Binder/NameofBinder.cs

```diff
@@ -2,19 +2,65 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
+    /// <summary>
+    /// If a proper method named "nameof" exists in the outer scopes, <see cref="IsNameofOperator"/> is false and this binder does nothing.
+    /// Otherwise, it relaxes the instance-vs-static requirement for top-level member access expressions
+    /// and when inside an attribute on a method it adds type parameters from the target of that attribute.
+    /// To do so, it works together with <see cref="ContextualAttributeBinder"/>.
+    /// 
+    /// For other attributes (on types, type parameters or parameters) we use a WithTypeParameterBinder directly
+    /// in the binder chain and some filtering (<see cref="LookupOptions.MustNotBeMethodTypeParameter"/>) to keep
+    /// pre-existing behavior.
+    /// </summary>
     internal sealed class NameofBinder : Binder
     {
         private readonly SyntaxNode _nameofArgument;
+        private readonly WithTypeParametersBinder? _withTypeParametersBinder;
+        private ThreeState _lazyIsNameofOperator;
 
-        public NameofBinder(SyntaxNode nameofArgument, Binder next) : base(next)
+        internal NameofBinder(SyntaxNode nameofArgument, Binder next, WithTypeParametersBinder? withTypeParametersBinder)
+            : base(next)
         {
             _nameofArgument = nameofArgument;
+            _withTypeParametersBinder = withTypeParametersBinder;
         }
 
-        protected override SyntaxNode EnclosingNameofArgument => _nameofArgument;
+        private bool IsNameofOperator
+        {
+            get
+            {
+                if (!_lazyIsNameofOperator.HasValue())
+                {
+                    _lazyIsNameofOperator = ThreeStateHelpers.ToThreeState(!NextRequired.InvocableNameofInScope());
+                }
+
+                return _lazyIsNameofOperator.Value();
+            }
+        }
+
+        internal override bool IsInsideNameof => IsNameofOperator || base.IsInsideNameof;
+
+        protected override SyntaxNode? EnclosingNameofArgument => IsNameofOperator ? _nameofArgument : base.EnclosingNameofArgument;
```

Comment:
It looks like this implementation might be important in a nested scenario of some kind? Basically, I wasn't sure if it matters that we delegate to the base here when this is not a nameof operator, versus just returning null when this is not a nameof operator.

Reply:
Yes, a nameof inside a nameof.